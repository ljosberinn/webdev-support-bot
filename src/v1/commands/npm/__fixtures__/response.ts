export const response = [
  {
    date: '2020-03-19T19:53:13.309Z',
    description: 'React is a JavaScript library for building user interfaces.',
    keywords: ['react'],
    links: {
      bugs: 'https://github.com/facebook/react/issues',
      homepage: 'https://reactjs.org/',
      npm: 'https://www.npmjs.com/package/react',
      repository: 'https://github.com/facebook/react',
    },
    maintainers: [
      {
        email: 'npm@andrewclark.io',
        username: 'acdlite',
      },
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'lunaris.ruan@gmail.com',
        username: 'lunaruan',
      },
      {
        email: 'sebastian@calyptus.eu',
        username: 'sebmarkbage',
      },
      {
        email: 'npm@sophiebits.com',
        username: 'sophiebits',
      },
      {
        email: 'dg@domgan.com',
        username: 'trueadm',
      },
    ],
    name: 'react',
    publisher: {
      email: 'npm@andrewclark.io',
      username: 'acdlite',
    },
    scope: 'unscoped',
    version: '16.13.1',
  },
  {
    author: {
      name: 'Jed Watson',
    },
    date: '2020-03-22T04:02:30.633Z',
    description: 'A Select control built with and for ReactJS',
    keywords: [
      'combobox',
      'form',
      'input',
      'multiselect',
      'react',
      'react-component',
      'select',
      'ui',
    ],
    links: {
      bugs: 'https://github.com/JedWatson/react-select/issues',
      homepage: 'https://github.com/JedWatson/react-select#readme',
      npm: 'https://www.npmjs.com/package/react-select',
      repository: 'https://github.com/JedWatson/react-select',
    },
    maintainers: [
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'cc.lee@live.com.au',
        username: 'gwyneplaine',
      },
      {
        email: 'jed.watson@me.com',
        username: 'jedwatson',
      },
      {
        email: 'mitchell@mitchellhamilton.me',
        username: 'mitchellhamilton',
      },
      {
        email: 'automation+react-select@thinkmill.com.au',
        username: 'react-select-release-bot',
      },
    ],
    name: 'react-select',
    publisher: {
      email: 'automation+react-select@thinkmill.com.au',
      username: 'react-select-release-bot',
    },
    scope: 'unscoped',
    version: '3.1.0',
  },
  {
    author: {
      name: 'Matt Zabriskie',
    },
    date: '2020-05-14T14:38:16.281Z',
    description: 'React draggable component',
    keywords: ['react', 'draggable', 'react-component'],
    links: {
      bugs: 'https://github.com/mzabriskie/react-draggable/issues',
      homepage: 'https://github.com/mzabriskie/react-draggable',
      npm: 'https://www.npmjs.com/package/react-draggable',
      repository: 'https://github.com/mzabriskie/react-draggable',
    },
    maintainers: [
      {
        email: 'mzabriskie@gmail.com',
        username: 'mzabriskie',
      },
      {
        email: 'samuel.trace.reed@gmail.com',
        username: 'strml',
      },
    ],
    name: 'react-draggable',
    publisher: {
      email: 'samuel.trace.reed@gmail.com',
      username: 'strml',
    },
    scope: 'unscoped',
    version: '4.4.2',
  },
  {
    author: {
      email: 'travis@souporserious.com',
      name: 'Travis Arnold',
      url: 'http://souporserious.com',
    },
    date: '2020-04-28T09:55:08.691Z',
    description: 'Official library to use Popper on React projects',
    keywords: [
      'react',
      'react-popper',
      'popperjs',
      'component',
      'drop',
      'tooltip',
      'popover',
    ],
    links: {
      bugs: 'https://github.com/popperjs/react-popper/issues',
      homepage: 'https://github.com/popperjs/react-popper',
      npm: 'https://www.npmjs.com/package/react-popper',
      repository: 'https://github.com/popperjs/react-popper',
    },
    maintainers: [
      {
        email: 'federico.zivolo@gmail.com',
        username: 'fezvrasta',
      },
      {
        email: 'ftntravis@gmail.com',
        username: 'souporserious',
      },
    ],
    name: 'react-popper',
    publisher: {
      email: 'federico.zivolo@gmail.com',
      username: 'fezvrasta',
    },
    scope: 'unscoped',
    version: '2.2.3',
  },
  {
    author: {
      name: 'Param Aggarwal',
    },
    date: '2020-04-26T08:29:08.419Z',
    description: 'Simple HTML5 drag-drop zone with React.js',
    keywords: ['react-component', 'react', 'drag', 'drop', 'upload'],
    links: {
      bugs: 'https://github.com/react-dropzone/react-dropzone/issues',
      homepage: 'https://github.com/react-dropzone/react-dropzone',
      npm: 'https://www.npmjs.com/package/react-dropzone',
      repository: 'https://github.com/react-dropzone/react-dropzone',
    },
    maintainers: [
      {
        email: 'andrey@okonet.ru',
        username: 'okonet',
      },
      {
        email: 'rolandjitsu@gmail.com',
        username: 'rolandjitsu',
      },
      {
        email: 'rmarkins@gmail.com',
        username: 'rxmarbles',
      },
    ],
    name: 'react-dropzone',
    publisher: {
      email: 'rolandjitsu@gmail.com',
      username: 'rolandjitsu',
    },
    scope: 'unscoped',
    version: '11.0.1',
  },
  {
    author: {
      email: 'dan.abramov@me.com',
      name: 'Dan Abramov',
      url: 'https://github.com/gaearon',
    },
    date: '2020-02-18T03:43:52.869Z',
    description: 'Official React bindings for Redux',
    keywords: ['react', 'reactjs', 'redux'],
    links: {
      bugs: 'https://github.com/reduxjs/react-redux/issues',
      homepage: 'https://github.com/reduxjs/react-redux',
      npm: 'https://www.npmjs.com/package/react-redux',
      repository: 'https://github.com/reduxjs/react-redux',
    },
    maintainers: [
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'timdorr@timdorr.com',
        username: 'timdorr',
      },
    ],
    name: 'react-redux',
    publisher: {
      email: 'mark.erikson@gmail.com',
      username: 'acemarke',
    },
    scope: 'unscoped',
    version: '7.2.0',
  },
  {
    author: {
      email: 'case@casesandberg.com',
      name: 'case',
      username: 'case',
    },
    date: '2020-05-04T00:55:19.458Z',
    description:
      'A Collection of Color Pickers from Sketch, Photoshop, Chrome & more',
    keywords: [
      'react',
      'color picker',
      'react-component',
      'colorpicker',
      'picker',
      'sketch',
      'chrome',
      'photoshop',
      'material design',
      'popup',
    ],
    links: {
      homepage: 'http://casesandberg.github.io/react-color/',
      npm: 'https://www.npmjs.com/package/react-color',
    },
    maintainers: [
      {
        email: 'case@casesandberg.com',
        username: 'case',
      },
    ],
    name: 'react-color',
    publisher: {
      email: 'case@casesandberg.com',
      username: 'case',
    },
    scope: 'unscoped',
    version: '2.18.1',
  },
  {
    date: '2020-03-19T19:53:23.004Z',
    description: 'React package for working with the DOM.',
    keywords: ['react'],
    links: {
      bugs: 'https://github.com/facebook/react/issues',
      homepage: 'https://reactjs.org/',
      npm: 'https://www.npmjs.com/package/react-dom',
      repository: 'https://github.com/facebook/react',
    },
    maintainers: [
      {
        email: 'npm@andrewclark.io',
        username: 'acdlite',
      },
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'lunaris.ruan@gmail.com',
        username: 'lunaruan',
      },
      {
        email: 'sebastian@calyptus.eu',
        username: 'sebmarkbage',
      },
      {
        email: 'npm@sophiebits.com',
        username: 'sophiebits',
      },
      {
        email: 'dg@domgan.com',
        username: 'trueadm',
      },
      {
        email: 'paul@oshannessy.com',
        username: 'zpao',
      },
    ],
    name: 'react-dom',
    publisher: {
      email: 'npm@andrewclark.io',
      username: 'acdlite',
    },
    scope: 'unscoped',
    version: '16.13.1',
  },
  {
    author: {
      name: 'Felix Kling',
    },
    date: '2020-02-26T00:42:09.515Z',
    description:
      'A CLI and toolkit to extract information from React components for documentation generation.',
    keywords: ['react', 'documentation-generation'],
    links: {
      bugs: 'https://github.com/reactjs/react-docgen/issues',
      homepage: 'https://github.com/reactjs/react-docgen#readme',
      npm: 'https://www.npmjs.com/package/react-docgen',
      repository: 'https://github.com/reactjs/react-docgen',
    },
    maintainers: [
      {
        email: 'daniel@tschinder.de',
        username: 'danez',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
    ],
    name: 'react-docgen',
    publisher: {
      email: 'daniel@tschinder.de',
      username: 'danez',
    },
    scope: 'unscoped',
    version: '5.3.0',
  },
  {
    date: '2020-05-15T21:38:04.997Z',
    description:
      'Hooks for building lightweight, fast and extendable datagrids for React',
    keywords: ['react', 'table', 'react-table', 'datagrid'],
    links: {
      bugs: 'https://github.com/tannerlinsley/react-table/issues',
      homepage: 'https://react-table.js.org',
      npm: 'https://www.npmjs.com/package/react-table',
      repository: 'https://github.com/tannerlinsley/react-table',
    },
    maintainers: [
      {
        email: 'tannerlinsley@gmail.com',
        username: 'tannerlinsley',
      },
      {
        email: 'zach.h.sherman@gmail.com',
        username: 'zsherman',
      },
    ],
    name: 'react-table',
    publisher: {
      email: 'tannerlinsley@gmail.com',
      username: 'tannerlinsley',
    },
    scope: 'unscoped',
    version: '7.1.0',
  },
];
