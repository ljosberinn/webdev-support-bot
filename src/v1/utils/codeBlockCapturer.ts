const BACKTICKS = '```';

export type CodeBlockData = {
  code: string;
  language: string;
};

export function createCodeBlockCapturer(langs: string[] = []) {
  const langAlts = langs.join('|');
  return function* (str: string): Generator<CodeBlockData> {
    const langRegex = new RegExp(
      String.raw`${BACKTICKS}(?<language>${langAlts})\n(?<code>[\s\S]+?)\n${BACKTICKS}`,
      'gi'
    );
    const matches = str.matchAll(langRegex);
    for (const {
      groups: { language, code },
    } of matches) {
      yield {
        language: language || '',
        code: code || '',
      };
    }
  };
}
