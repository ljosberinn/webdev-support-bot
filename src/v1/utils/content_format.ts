import { Message } from 'discord.js';

const linebreakPattern = /\n/gim;

export const generateCleanContent = (msg: Message) =>
  msg.cleanContent.replace(linebreakPattern, ' ').toLowerCase();

export const stripMarkdownQuote = (msg: string) => msg.replace(/^> .+$/gm, '');
