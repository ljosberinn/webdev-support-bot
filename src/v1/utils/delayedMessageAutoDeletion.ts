import { Message } from 'discord.js';

import { missingRightsDeletion } from './errors';

const THIRTY_SECONDS_IN_MS = 30 * 1000;

export const delayedMessageAutoDeletion = (
  msg: Message,
  timeout = THIRTY_SECONDS_IN_MS
) => {
  // required so tests on CI dont crash
  if (msg) {
    setTimeout(async () => {
      try {
        await msg.delete();
      } catch (error) {
        // eslint-disable-next-line no-console
        console.warn("Couldn't delete message", error);

        try {
          await msg.edit(missingRightsDeletion);
        } catch {
          // eslint-disable-next-line no-console
          console.info(
            "Couldn't edit message after trying to delete, probably removed by someone else."
          );
        }
      }
    }, timeout);
  }

  return;
};
