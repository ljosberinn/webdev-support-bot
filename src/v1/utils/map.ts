/**
 * same as Array#map but works on all iterables, returns an iterable
 *
 * @export
 * @template T input iterable
 * @template U output iterable
 * @param {(item: T, index: number, iter: Iterable<T>) => U} fn
 * @param {Iterable<T>} iter
 */
export function* mapʹ<T, U>(
  fn: (item: T, index: number, iter: Iterable<T>) => U,
  iter: Iterable<T>
) {
  let i = 0;
  for (const item of iter) {
    yield fn(item, i++, iter);
  }
}

export function map<T, U>(
  fn: (item: T, index: number, iter: Iterable<T>) => U
) {
  return (iter: Iterable<T>) => mapʹ(fn, iter);
}
