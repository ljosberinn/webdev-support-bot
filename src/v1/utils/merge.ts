export function* merge<T>(...iterables: Iterable<T>[]) {
  for (const iterable of iterables) {
    yield* iterable;
  }
}
