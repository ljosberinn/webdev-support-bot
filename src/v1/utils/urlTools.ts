import { Message } from 'discord.js';
import { HeadersInit } from 'node-fetch';

import { delayedMessageAutoDeletion } from './delayedMessageAutoDeletion';
import { Provider } from './discordTools';
import { noResults, invalidResponse } from './errors';
import useData from './useData';

const SEARCH_TERM = '%SEARCH%';
const TERM = '%TERM%';

type ProviderMap = {
  [key in Provider]: {
    search: string;
    color: number;
    createTitle: (term: string) => string;
    icon: string;
    help: string;
    direct?: string;
    getExtendedInfoUrl?: (pkg: string) => string;
  };
};

export const providers: ProviderMap = {
  bundlephobia: {
    color: 0xffffff,
    createTitle: (searchTerm: string) =>
      `Bundlephobia results for *${searchTerm}*`,
    direct: `https://bundlephobia.com/result?p=${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://bundlephobia.com/api/size?package=${pkg}&record=true`,
    help: '!bundlephobia @chakra-ui/core',
    icon: 'https://bundlephobia.com/android-chrome-192x192.png',
    search: `https://api.npms.io/v2/search/suggestions?q=${SEARCH_TERM}`,
  },
  caniuse: {
    color: 0xdb5600,
    createTitle: (searchTerm: string) => `CanIUse results for *${searchTerm}*`,
    direct: `https://caniuse.com/#feat=${TERM}`,
    getExtendedInfoUrl: (text: string) =>
      `https://caniuse.com/process/get_feat_data.php?type=support-data&feat=${text}`,
    help: '!caniuse IntersectionObserver',
    icon: 'https://caniuse.com/img/favicon-128.png',
    search: `https://caniuse.com/process/query.php?search=${SEARCH_TERM}`,
  },
  composer: {
    color: 0xf28d1a,
    createTitle: (searchTerm: string) => `Packagist results for ${searchTerm}`,
    direct: `https://packagist.org/packages/${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://packagist.org/packages/${pkg}.json`,
    help: '!composer sentry',
    icon: 'https://packagist.org/bundles/packagistweb/img/logo-small.png',
    search: `https://packagist.org/search.json?q=${SEARCH_TERM}`,
  },
  github: {
    color: 0x24292e,
    createTitle: (searchTerm: string) => `GitHub results for *${searchTerm}*`,
    direct: `https://github.com/${TERM}`,
    help: '!github react',
    icon:
      'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
    search: `https://api.github.com/search/repositories?q=${SEARCH_TERM}`,
  },
  mdn: {
    color: 0x83d0f2,
    createTitle: (searchTerm: string) => `MDN results for *${searchTerm}*`,
    direct: `https://developer.mozilla.org${TERM}`,
    help: '!mdn localStorage',
    icon: 'https://avatars0.githubusercontent.com/u/7565578',
    search: `https://developer.mozilla.org/api/v1/search?q=${SEARCH_TERM}&locale=en-US`,
  },
  npm: {
    color: 0xfb3e44,
    createTitle: (searchTerm: string) => `NPM results for *${searchTerm}*`,
    help: '!npm react',
    icon: 'https://avatars0.githubusercontent.com/u/6078720',
    search: `https://www.npmjs.com/search/suggestions?q=${SEARCH_TERM}`,
  },
  php: {
    color: 0x8892bf,
    createTitle: (searchTerm: string) => `PHP.net results for *${searchTerm}*`,
    direct: `https://www.php.net/${TERM}`,
    help: '!php echo',
    icon: 'https://www.php.net/images/logos/php-logo.svg',
    search: `https://www.php.net/${SEARCH_TERM}`,
  },
};

export const HELP_KEYWORD = '--help';
export const FORMATTING_KEYWORD = '!formatting';
export const FORMATTING_KEYWORD_ALT = '!format';
export const CODE_KEYWORD = '!code';
export const VSCODE_KEYWORD = '!vscode';
export const JOB_POSTING_KEYWORD = '!post';
export const JQUERY_KEYWORD = '!jquery';
export const POINTS_KEYWORD = '!points';
export const LEADERBOARD_KEYWORD = '!leaderboard';
export const DECAY_KEYWORD = '!decay';
export const MODULE_KEYWORD = '!modules';
export const LOCKFILE_KEYWORD = '!lockfile';
export const FLEXBOX_KEYWORD = '!flexbox';

/**
 * dynamic regExp matching all possible Object.keys(providers) as keyword
 */
export const KEYWORD_REGEXP = new RegExp(
  '^!(%PLACEHOLDER%)\\s+'.replace(
    '%PLACEHOLDER%',
    Object.keys(providers)
      .reduce((carry, keyword) => [...carry, keyword], [])
      .join('|')
  ),
  'i'
);

export const getSearchUrl = (provider: Provider, search: string) => {
  if (providers[provider]) {
    return providers[provider].search.replace(SEARCH_TERM, encodeURI(search));
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const buildDirectUrl = (provider: Provider, href: string) => {
  if (providers[provider]) {
    return providers[provider].direct.replace(TERM, href);
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const getExtendedInfoUrl = (provider: Provider, term: string) => {
  if (providers[provider] && providers[provider].getExtendedInfoUrl) {
    return providers[provider].getExtendedInfoUrl(term);
  }

  throw new Error(
    `provider or provider.getExtendedInfoUrl not implemented at provider: ${provider}`
  );
};

interface GetDataParams {
  msg: Message;
  provider: Provider;
  searchTerm: string;
  sanitizeData?: (data: any) => Partial<any>;
  isInvalidData: (data: any) => boolean;
  headers?: HeadersInit;
}

export const getData = async <T>({
  msg,
  provider,
  searchTerm,
  sanitizeData,
  isInvalidData,
  headers,
}: GetDataParams): Promise<Partial<T>> => {
  const searchUrl = getSearchUrl(provider, searchTerm);
  const { error, json: data } = await useData<T>(searchUrl, 'json', headers);

  if (error) {
    await msg.reply(invalidResponse);
    return;
  }

  const sanitizedData = sanitizeData ? sanitizeData(data) : data;

  if (isInvalidData(sanitizedData)) {
    const sentMessage = await msg.reply(noResults(searchTerm));

    delayedMessageAutoDeletion(sentMessage);
    return;
  }

  return sanitizedData;
};
