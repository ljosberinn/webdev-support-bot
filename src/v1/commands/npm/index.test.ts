import { getChosenResult } from '../../utils/discordTools';
import { getData } from '../../utils/urlTools';
import { response } from './__fixtures__/response';

import { buildNPMQueryHandler } from '.';

describe('npm', () => {
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    delete: replyMock,
    reply: replyMock,
  };

  const fetch: jest.MockedFunction<typeof getData> = jest.fn();
  const choose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  afterEach(() => jest.resetAllMocks());

  test('returns when request fails', async () => {
    fetch.mockResolvedValue(null);
    const handler = buildNPMQueryHandler(fetch);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'npm',
      searchTerm: 'search term',
    });

    expect(choose).not.toBeCalled();
  });

  test('awaits user response when request works', async () => {
    const editMock = jest.fn();
    fetch.mockResolvedValue(response);
    choose.mockResolvedValue({
      author: { name: '', url: '' },
      description: '',
      externalUrls: { homepage: '', repository: '' },
      lastUpdate: '',
      maintainers: '10',
      name: 'React',
      url: 'http://react',
    });

    msg.channel.send.mockResolvedValue({ edit: editMock });

    const handler = buildNPMQueryHandler(fetch, choose, () => '20 days');
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'npm',
      searchTerm: 'search term',
    });

    expect(msg.channel.send.mock.calls[0]).toMatchSnapshot();
    expect(choose.mock.calls[0]).toMatchSnapshot();
    expect(sendMock.mock.calls[0]).toMatchSnapshot();
  });
});
