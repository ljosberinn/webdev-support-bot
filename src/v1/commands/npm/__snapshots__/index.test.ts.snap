// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm awaits user response when request works 1`] = `
Array [
  Object {
    "embed": Object {
      "author": null,
      "color": 16465476,
      "description": "1. [**react** - *React is a JavaScript library for building user interfaces.*](https://www.npmjs.com/package/react)
2. [**react-select** - *A Select control built with and for ReactJS*](https://www.npmjs.com/package/react-select)
3. [**react-draggable** - *React draggable component*](https://www.npmjs.com/package/react-draggable)
4. [**react-popper** - *Official library to use Popper on React projects*](https://www.npmjs.com/package/react-popper)
5. [**react-dropzone** - *Simple HTML5 drag-drop zone with React.js*](https://www.npmjs.com/package/react-dropzone)
6. [**react-redux** - *Official React bindings for Redux*](https://www.npmjs.com/package/react-redux)
7. [**react-color** - *A Collection of Color Pickers from Sketch, Photoshop,...*](https://www.npmjs.com/package/react-color)
8. [**react-dom** - *React package for working with the DOM.*](https://www.npmjs.com/package/react-dom)
9. [**react-docgen** - *A CLI and toolkit to extract information from React...*](https://www.npmjs.com/package/react-docgen)
10. [**react-table** - *Hooks for building lightweight, fast and extendable...*](https://www.npmjs.com/package/react-table)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
      "fields": Array [],
      "footer": Object {
        "iconURL": "https://avatars0.githubusercontent.com/u/6078720",
        "text": "at least 10 packages found",
      },
      "title": "NPM results for *search term*",
      "url": "https://npmjs.com/search?q=search%20term",
    },
  },
]
`;

exports[`npm awaits user response when request works 2`] = `
Array [
  Object {
    "edit": [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "embed": Object {
              "author": Object {
                "name": "",
                "url": "",
              },
              "color": 16465476,
              "description": "",
              "fields": Array [
                Object {
                  "inline": false,
                  "name": "add to your project",
                  "value": "\`\`\`bash
npm install React
yarn add React
\`\`\`",
                },
                Object {
                  "inline": true,
                  "name": ":writing_hand: maintainers",
                  "value": "10",
                },
              ],
              "footer": Object {
                "iconURL": "https://avatars0.githubusercontent.com/u/6078720",
                "text": "last updated ",
              },
              "title": "React",
              "url": "http://react",
            },
          },
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  Object {
    "channel": Object {
      "send": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "embed": Object {
                "author": null,
                "color": 16465476,
                "description": "1. [**react** - *React is a JavaScript library for building user interfaces.*](https://www.npmjs.com/package/react)
2. [**react-select** - *A Select control built with and for ReactJS*](https://www.npmjs.com/package/react-select)
3. [**react-draggable** - *React draggable component*](https://www.npmjs.com/package/react-draggable)
4. [**react-popper** - *Official library to use Popper on React projects*](https://www.npmjs.com/package/react-popper)
5. [**react-dropzone** - *Simple HTML5 drag-drop zone with React.js*](https://www.npmjs.com/package/react-dropzone)
6. [**react-redux** - *Official React bindings for Redux*](https://www.npmjs.com/package/react-redux)
7. [**react-color** - *A Collection of Color Pickers from Sketch, Photoshop,...*](https://www.npmjs.com/package/react-color)
8. [**react-dom** - *React package for working with the DOM.*](https://www.npmjs.com/package/react-dom)
9. [**react-docgen** - *A CLI and toolkit to extract information from React...*](https://www.npmjs.com/package/react-docgen)
10. [**react-table** - *Hooks for building lightweight, fast and extendable...*](https://www.npmjs.com/package/react-table)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
                "fields": Array [],
                "footer": Object {
                  "iconURL": "https://avatars0.githubusercontent.com/u/6078720",
                  "text": "at least 10 packages found",
                },
                "title": "NPM results for *search term*",
                "url": "https://npmjs.com/search?q=search%20term",
              },
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
    "delete": [MockFunction],
    "reply": [MockFunction],
  },
  Array [
    Object {
      "author": Object {
        "name": "acdlite",
        "url": "https://www.npmjs.com/~acdlite",
      },
      "description": "React is a JavaScript library for building user interfaces.",
      "externalUrls": Object {
        "homepage": "https://reactjs.org/",
        "repository": "https://github.com/facebook/react",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 8,
      "name": "react",
      "url": "https://www.npmjs.com/package/react",
    },
    Object {
      "author": Object {
        "name": "react-select-release-bot",
        "url": "https://www.npmjs.com/~react-select-release-bot",
      },
      "description": "A Select control built with and for ReactJS",
      "externalUrls": Object {
        "homepage": "https://github.com/JedWatson/react-select#readme",
        "repository": "https://github.com/JedWatson/react-select",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 5,
      "name": "react-select",
      "url": "https://www.npmjs.com/package/react-select",
    },
    Object {
      "author": Object {
        "name": "strml",
        "url": "https://www.npmjs.com/~strml",
      },
      "description": "React draggable component",
      "externalUrls": Object {
        "homepage": "https://github.com/mzabriskie/react-draggable",
        "repository": "https://github.com/mzabriskie/react-draggable",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 2,
      "name": "react-draggable",
      "url": "https://www.npmjs.com/package/react-draggable",
    },
    Object {
      "author": Object {
        "name": "fezvrasta",
        "url": "https://www.npmjs.com/~fezvrasta",
      },
      "description": "Official library to use Popper on React projects",
      "externalUrls": Object {
        "homepage": "https://github.com/popperjs/react-popper",
        "repository": "https://github.com/popperjs/react-popper",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 2,
      "name": "react-popper",
      "url": "https://www.npmjs.com/package/react-popper",
    },
    Object {
      "author": Object {
        "name": "rolandjitsu",
        "url": "https://www.npmjs.com/~rolandjitsu",
      },
      "description": "Simple HTML5 drag-drop zone with React.js",
      "externalUrls": Object {
        "homepage": "https://github.com/react-dropzone/react-dropzone",
        "repository": "https://github.com/react-dropzone/react-dropzone",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 3,
      "name": "react-dropzone",
      "url": "https://www.npmjs.com/package/react-dropzone",
    },
    Object {
      "author": Object {
        "name": "acemarke",
        "url": "https://www.npmjs.com/~acemarke",
      },
      "description": "Official React bindings for Redux",
      "externalUrls": Object {
        "homepage": "https://github.com/reduxjs/react-redux",
        "repository": "https://github.com/reduxjs/react-redux",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 2,
      "name": "react-redux",
      "url": "https://www.npmjs.com/package/react-redux",
    },
    Object {
      "author": Object {
        "name": "case",
        "url": "https://www.npmjs.com/~case",
      },
      "description": "A Collection of Color Pickers from Sketch, Photoshop, Chrome & more",
      "externalUrls": Object {
        "homepage": "http://casesandberg.github.io/react-color/",
        "repository": undefined,
      },
      "lastUpdate": "20 days ago",
      "maintainers": 1,
      "name": "react-color",
      "url": "https://www.npmjs.com/package/react-color",
    },
    Object {
      "author": Object {
        "name": "acdlite",
        "url": "https://www.npmjs.com/~acdlite",
      },
      "description": "React package for working with the DOM.",
      "externalUrls": Object {
        "homepage": "https://reactjs.org/",
        "repository": "https://github.com/facebook/react",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 9,
      "name": "react-dom",
      "url": "https://www.npmjs.com/package/react-dom",
    },
    Object {
      "author": Object {
        "name": "danez",
        "url": "https://www.npmjs.com/~danez",
      },
      "description": "A CLI and toolkit to extract information from React components for documentation generation.",
      "externalUrls": Object {
        "homepage": "https://github.com/reactjs/react-docgen#readme",
        "repository": "https://github.com/reactjs/react-docgen",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 2,
      "name": "react-docgen",
      "url": "https://www.npmjs.com/package/react-docgen",
    },
    Object {
      "author": Object {
        "name": "tannerlinsley",
        "url": "https://www.npmjs.com/~tannerlinsley",
      },
      "description": "Hooks for building lightweight, fast and extendable datagrids for React",
      "externalUrls": Object {
        "homepage": "https://react-table.js.org",
        "repository": "https://github.com/tannerlinsley/react-table",
      },
      "lastUpdate": "20 days ago",
      "maintainers": 2,
      "name": "react-table",
      "url": "https://www.npmjs.com/package/react-table",
    },
  ],
]
`;

exports[`npm awaits user response when request works 3`] = `
Array [
  Object {
    "embed": Object {
      "author": null,
      "color": 16465476,
      "description": "1. [**react** - *React is a JavaScript library for building user interfaces.*](https://www.npmjs.com/package/react)
2. [**react-select** - *A Select control built with and for ReactJS*](https://www.npmjs.com/package/react-select)
3. [**react-draggable** - *React draggable component*](https://www.npmjs.com/package/react-draggable)
4. [**react-popper** - *Official library to use Popper on React projects*](https://www.npmjs.com/package/react-popper)
5. [**react-dropzone** - *Simple HTML5 drag-drop zone with React.js*](https://www.npmjs.com/package/react-dropzone)
6. [**react-redux** - *Official React bindings for Redux*](https://www.npmjs.com/package/react-redux)
7. [**react-color** - *A Collection of Color Pickers from Sketch, Photoshop,...*](https://www.npmjs.com/package/react-color)
8. [**react-dom** - *React package for working with the DOM.*](https://www.npmjs.com/package/react-dom)
9. [**react-docgen** - *A CLI and toolkit to extract information from React...*](https://www.npmjs.com/package/react-docgen)
10. [**react-table** - *Hooks for building lightweight, fast and extendable...*](https://www.npmjs.com/package/react-table)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
      "fields": Array [],
      "footer": Object {
        "iconURL": "https://avatars0.githubusercontent.com/u/6078720",
        "text": "at least 10 packages found",
      },
      "title": "NPM results for *search term*",
      "url": "https://npmjs.com/search?q=search%20term",
    },
  },
]
`;
